src/palindrom-network-channel.js

```diff
@@ -0,0 +1,355 @@
+import WebSocket from 'websocket';
+import nodeFetch from 'node-fetch';
+
+const CLIENT = 'Client';
+const SERVER = 'Server';
+
+/**
+ * Replaces http and https to ws and wss in a URL and returns it as a string.
+ * @param  {String} remoteUrl HTTP remote address
+ * @return {String}           WS address
+ */
+function toWebSocketURL(remoteUrl) {
+    /* replace 'http' strictly in the beginning of the string,
+    this covers http and https */
+    return remoteUrl.replace(/^http/i, 'ws');
+}
+
+export default class PalindromNetworkChannel {
+    constructor(
+        palindrom,
+        remoteUrl,
+        useWebSocket,
+        onReceive,
+        onSend,
+        onConnectionError,
+        onSocketOpened,
+        onFatalError,
+        onStateChange
+    ) {
+        // TODO(tomalec): to be removed once we will achieve better separation of concerns
+        this.palindrom = palindrom;
+
+        if (typeof window !== 'undefined' && window.location) {
+            this.remoteUrl = new URL(remoteUrl, window.location.href);
+        } else {
+            // in Node, URL is absolute
+            this.remoteUrl = new URL(remoteUrl);
+        }
+
+        onReceive && (this.onReceive = onReceive);
+        onSend && (this.onSend = onSend);
+        onConnectionError && (this.onConnectionError = onConnectionError);
+        onFatalError && (this.onFatalError = onFatalError);
+        onStateChange && (this.onStateChange = onStateChange);
+        onSocketOpened && (this.onSocketOpened = onSocketOpened);
+
+        Object.defineProperty(this, 'useWebSocket', {
+            get: function() {
+                return useWebSocket;
+            },
+            set: newValue => {
+                useWebSocket = newValue;
+
+                if (newValue == false) {
+                    if (this._ws) {
+                        this._ws.onclose = function() {
+                            //overwrites the previous onclose
+                            this._ws = null;
+                        };
+                        this._ws.close();
+                    }
+                    // define wsUrl if needed
+                } else if (!this.wsUrl) {
+                    this.wsUrl = toWebSocketURL(this.remoteUrl.href);
+                }
+                return useWebSocket;
+            }
+        });
+    }
+    get useWebSocket() {
+        return this._useWebSocket;
+    }
+    set useWebSocket(newValue) {
+        this._useWebSocket = newValue;
+
+        if (newValue == false) {
+            if (this._ws) {
+                this._ws.onclose = () => {
+                    //overwrites the previous onclose
+                    this._ws = null;
+                };
+                this._ws.close();
+            }
+            // define wsUrl if needed
+        } else if (!this.wsUrl) {
+            this.wsUrl = toWebSocketURL(this.remoteUrl.href);
+        }
+        return this.useWebSocket;
+    }
+
+    async _establish(reconnectionPendingData = null) {
```

Comment:
for NoQueue, and non-OT Queue, `reconnectionPendingData = palindrom.queue.pending === undefined`. So for those cases it will do regular establish connection. 

@alshakero is it by design?

@warpech should we bother about non-OT scenarios? or make Palindrom require OT compliant server? On one hand it will simplify maintenance, but would but more constrains on eventual servers other than Starcounter.

Reply:
> @warpech should we bother about non-OT scenarios? or make Palindrom require OT compliant server? On one hand it will simplify maintenance, but would but more constrains on eventual servers other than Starcounter.

I have added this question to my list of FAQs to be asked tomorrow.

My personal opinion is to only support the OT scenario and remove the code that makes OT optional. 

@miyconst what's your opinion?

Reply:
@miyconst wrote:

> we can keep [OT] required. Don't know a single use case where OT would not be needed.

Reply:
I created a new issue to track this: https://github.com/Palindrom/Palindrom/issues/241